apps/docs/content/guides/auth/oauth-server/token-security.mdx

```diff
@@ -0,0 +1,436 @@
+    const { data: agent } = await supabase
+      .from('approved_ai_agents')
+      .select('name, max_data_retention_days')
+      .eq('client_id', client_id)
+      .single()
+
+    customClaims.aud = `https://mcp.myapp.com/${client_id}`
+    customClaims.ai_agent = true
+    customClaims.agent_name = agent?.name
+    customClaims.max_retention = agent?.max_data_retention_days
+  }
+
+  return new Response(JSON.stringify({ claims: customClaims }), {
+    headers: { 'Content-Type': 'application/json' },
+  })
+})
+```
+
+Use these custom claims in RLS:
+
+```sql
+-- Policy based on custom claims
+CREATE POLICY "Read-only clients cannot modify"
+ON user_data FOR UPDATE
+USING (
+  auth.uid() = user_id AND
+  (auth.jwt() -> 'user_metadata' ->> 'read_only')::boolean IS NOT TRUE
+);
+
+-- Policy based on audience claim
+CREATE POLICY "Only specific audience can access"
+ON api_data FOR SELECT
+USING (
+  auth.uid() = user_id AND
+  (auth.jwt() ->> 'aud') IN (
+    'https://api.myapp.com',
+    'https://mobile.myapp.com'
+  )
+);
+```
+
+## Security best practices
+
+### 1. Principle of least privilege
+
+Grant OAuth clients only the minimum permissions they need:
+
+```sql
+-- Bad: Grant all access by default
+CREATE POLICY "OAuth clients full access"
+ON user_data FOR ALL
+USING (auth.uid() = user_id);
+
+-- Good: Grant specific access per client
+CREATE POLICY "Specific client specific access"
+ON user_data FOR SELECT
+USING (
+  auth.uid() = user_id AND
+  (auth.jwt() ->> 'client_id') = 'trusted-client-id'
+);
+```
+
+### 2. Separate policies for OAuth clients
+
+Create dedicated policies for OAuth clients rather than mixing them with user policies:
+
+```sql
+-- User access
+CREATE POLICY "Users access their own data"
+ON user_data FOR ALL
+USING (
+  auth.uid() = user_id AND
+  (auth.jwt() ->> 'client_id') IS NULL
+);
+
+-- OAuth client access (separate policy)
+CREATE POLICY "OAuth clients limited access"
+ON user_data FOR SELECT
+USING (
+  auth.uid() = user_id AND
+  (auth.jwt() ->> 'client_id') IN ('client-1', 'client-2')
+);
+```
+
+### 3. Regularly audit OAuth clients
+
+Track and review which clients have access:
+
+```sql
+-- View all active OAuth clients
+SELECT
+  oc.client_id,
+  oc.name,
+  oc.created_at,
+  COUNT(DISTINCT s.user_id) as active_users
+FROM auth.oauth_clients oc
+LEFT JOIN auth.sessions s ON s.client_id = oc.client_id
+WHERE s.created_at > NOW() - INTERVAL '30 days'
+GROUP BY oc.client_id, oc.name, oc.created_at;
+```
+
+## Testing your policies
+
+Always test your RLS policies before deploying to production:
+
+```sql
+-- Test as a specific OAuth client
+SET request.jwt.claims = '{
+  "sub": "test-user-uuid",
+  "role": "authenticated",
+  "client_id": "test-client-id"
+}';
+
+-- Test queries
+SELECT * FROM user_data WHERE user_id = 'test-user-uuid';
+
+-- Reset
+RESET request.jwt.claims;
+```
+
+Or use the Supabase Dashboard's [RLS policy tester](/dashboard/project/_/auth/policies).
+
+## Troubleshooting
```

Comment:
Same as before, might be better in troubleshooting section?

Reply:
We can for sure. The troubleshooting docs have something of a different format, but there is a template, so let me know if you need any help.